### PR TITLE
[SPARK-8020] Spark SQL in spark-defaults.conf make metadataHive get constructed too early

### PR DESCRIPTION
**Pull request by yhuai**
https://issues.apache.org/jira/browse/SPARK-8020

### CODE REVIEW

**Comment by yhuai** [comment]
cc @marmbrus

**Comment by marmbrus** [comment]
LGTM

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 1]
```diff
@@ -37,6 +38,48 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+          // caused by SPARK-8020. Other problems triggered by SPARK-8020
+          // (e.g. using Hive 0.13.1's metastore client to connect to the a 0.12 metastore)
+          // are not easy to test.
+          Array(
+            ("spark.sql.hive.metastore.version" -> metastoreVersion),
+            ("spark.sql.hive.metastore.jars" -> metastoreJars)) ++ others
+        }
+      }
+    sparkConf
+      .set("spark.sql.hive.metastore.version", "12")
```
Comment (jeanlyn) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Does `12` equate to `0.12.0` or `1.2`?

Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Yeah (https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala#L49).

**Pull request closed without merging**